mixnet/client.py

```diff
@@ -1,76 +1,122 @@
 from mixnet.node import PacketQueue
 from mixnet.packet import PacketBuilder
 from mixnet.poisson import poisson_interval_sec
+from mixnet.topology import MixnetTopology
 
 
-async def mixclient_emitter(
-    mixnet: Mixnet,
-    emission_rate_per_min: int,  # Poisson rate parameter: lambda in the spec
-    redundancy: int,  # b in the spec
-    real_packet_queue: PacketQueue,
-    outbound_socket: PacketQueue,
-):
-    """
-    Emit packets at the Poisson emission_rate_per_min.
-
-    If a real packet is scheduled to be sent, this thread sends the real packet to the mixnet,
-    and schedules redundant real packets to be emitted in the next turns.
-
-    If no real packet is not scheduled, this thread emits a cover packet according to the emission_rate_per_min.
-    """
-
-    redundant_real_packet_queue: PacketQueue = asyncio.Queue()
-
-    emission_notifier_queue = asyncio.Queue()
-    _ = asyncio.create_task(
-        emission_notifier(emission_rate_per_min, emission_notifier_queue)
-    )
-
-    while True:
-        # Wait until the next emission time
-        _ = await emission_notifier_queue.get()
-        try:
-            await emit(
-                mixnet,
-                redundancy,
-                real_packet_queue,
-                redundant_real_packet_queue,
-                outbound_socket,
-            )
-        finally:
-            # Python convention: indicate that the previously enqueued task has been processed
-            emission_notifier_queue.task_done()
-
-
-async def emit(
-    mixnet: Mixnet,
-    redundancy: int,  # b in the spec
-    real_packet_queue: PacketQueue,
-    redundant_real_packet_queue: PacketQueue,
-    outbound_socket: PacketQueue,
-):
-    if not redundant_real_packet_queue.empty():
-        addr, packet = redundant_real_packet_queue.get_nowait()
-        await outbound_socket.put((addr, packet))
-        return
-
-    if not real_packet_queue.empty():
-        addr, packet = real_packet_queue.get_nowait()
-        # Schedule redundant real packets
-        for _ in range(redundancy - 1):
-            redundant_real_packet_queue.put_nowait((addr, packet))
-            await outbound_socket.put((addr, packet))
-
-    packet, route = PacketBuilder.drop_cover(b"drop cover", mixnet).next()
-    await outbound_socket.put((route[0].addr, packet))
-
-
-async def emission_notifier(emission_rate_per_min: int, queue: asyncio.Queue):
-    while True:
-        await asyncio.sleep(poisson_interval_sec(emission_rate_per_min))
-        queue.put_nowait(None)
+class MixClient:
+    __topology: MixnetTopology
+
+    __real_packet_queue: PacketQueue
+    __outbound_socket: PacketQueue
+    __task: asyncio.Task  # A reference just to prevent task from being garbage collected
+
+    @classmethod
+    async def new(
+        cls,
+        initial_topology: MixnetTopology,
+        emission_rate_per_min: int,  # Poisson rate parameter: lambda in the spec
+        redundancy: int,  # b in the spec
```

Comment:
Just an idea, shouldn't we consider the `emission_rate_per_min` and `redundancy` as parameters that are provided by the robustness layer? If we are allowing the robustness layer to configure the topology and manage that configuration, would it be better to relay all mixnet configuration aspects to the robustness layer?

I just think that it might be better to maintain all mixnet configuration logic in one place and execution in another.

Reply:
Good point. I've postponed it because in v1, only topology is updated by the robustness layer. But, yes, it's much better to move all mixnet config logics to the robustness layer, since that's why we've introduced the robustness layer. Thank you for bringing this up. I just refactored it: c402968.